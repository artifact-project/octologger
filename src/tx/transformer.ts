import ts from 'typescript';
import * as path from 'path';

const CWD = process.cwd();

type TxAPI = {
	checker: ts.TypeChecker;
	context: ts.TransformationContext
	sourceFile: ts.SourceFile;
}

function visitNode(node: ts.Node, {checker, sourceFile}: TxAPI) {
	if (ts.isCallExpression(node)) {
		const {
			arguments:args,
			typeArguments,
			expression,
		} = node;
		const signature = checker.getResolvedSignature(node);
		const returnType = signature && checker.typeToString(signature?.getReturnType())

		// console.log(node.getText(), returnType);

		if (!returnType || !returnType.startsWith('OctoShell<') || !args.length) {
			return node;
		}
		
		const {line, character} = ts.getLineAndCharacterOfPosition(sourceFile, node.getStart());
		const metaArg: any = ts.factory.createIdentifier(JSON.stringify([
			path.relative(CWD, sourceFile.fileName),
			line + 1,
			character + 1,
		]));
		
		try {
			metaArg.parent = (args[0] as any).parent;
			(args as any).unshift(metaArg);

			return ts.factory.updateCallExpression(
				node,
				expression,
				typeArguments,
				args,
			);
		} catch (err) {
			console.log('[octologger.tx] Update Call Expression failed:', err);
		}
	}

	return node;
}

function visitNodeAndChildren(
	node: ts.Node,
	api: TxAPI
) {
	return ts.visitEachChild(
		visitNode(node, api),
		(childNode) => visitNodeAndChildren(childNode, api),
		api.context,
	);
}

export function transformer(checkerOrProgram?: ts.TypeChecker | ts.Program) {
	let checker = checkerOrProgram && 'getTypeChecker' in checkerOrProgram
		? checkerOrProgram.getTypeChecker()
		: checkerOrProgram
	;

	return (context: ts.TransformationContext) => {
		return (sourceFile: ts.SourceFile, skip?: boolean) => {
			if (sourceFile.isDeclarationFile || skip === true) {
				return sourceFile;
			}

			if (!checker) {
				const program = ts.createProgram({
					rootNames: ['./index.ts'], 
					options: context.getCompilerOptions(),
				});

				checker = program.getTypeChecker();
				// sourceFile = program.getSourceFiles().find(({fileName}) => {
				// 	return fileName === sourceFile.fileName
				// })!;
			}

			sourceFile.forEachChild((node) => {
				visitNodeAndChildren(node, {
					checker: checker!,
					context,
					sourceFile,
				});
			});

			return sourceFile;
		}
	}
}
