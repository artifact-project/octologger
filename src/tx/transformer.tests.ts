import ts from 'typescript';
import * as fs from 'fs';
import * as path from 'path';
import { transformer } from './transformer';

const compilerOptions: ts.CompilerOptions = {
    module: ts.ModuleKind.ESNext,
    target: ts.ScriptTarget.ESNext,
    jsx: ts.JsxEmit.Preserve,
    removeComments: false,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
};

function readCompilerOptions(file: string): ts.CompilerOptions {
	const configFile = ts.findConfigFile(file, fs.existsSync);

	if (configFile) {
		const configSourceFile = ts.readJsonConfigFile(configFile, (f) => `${fs.readFileSync(f)}`);
		const config = ts.parseJsonSourceFileConfigFileContent(configSourceFile, ts.sys, path.dirname(configFile));

		return config.options;
	}

	return {};
}

const fixtureName = 'transformer.fixture.ts';
const fixtureFile = path.join(__dirname, fixtureName);
const fixtureContent = fs.readFileSync(fixtureFile) + '';
const tsCompilerOptions = {
	...readCompilerOptions(fixtureFile),
	noEmit: true,
	target: ts.ScriptTarget.ES2020,
	module: ts.ModuleKind.ES2020,
};


describe('transformer', () => {
	it('createProgram', () => {
		const program = ts.createProgram([fixtureFile], tsCompilerOptions);
		const checker = program.getTypeChecker();
		const context: ts.TransformationContext = {
			...(ts as any).nullTransformationContext as ts.TransformationContext,
			getCompilerOptions: () => program.getCompilerOptions(),
		};
		const tx = transformer(checker);
	
		for (const sourceFile of program.getSourceFiles()) {
			const isFixture = sourceFile.fileName.includes(fixtureName);
			const result = tx(context)(sourceFile, !isFixture);
			
			if (isFixture) {
				const printer = ts.createPrinter({ newLine: ts.NewLineKind.LineFeed });
				const out = printer.printFile(result);
				
				expect(`${sourceFile.fileName}\n\n${out}\n`).toMatchSnapshot();
			}
		}
	});

	xit('transpile', () => {
		const result = ts.transpile(fixtureContent, tsCompilerOptions);
		expect(result).toMatchSnapshot();
	});

	xit('transpileModule', () => {
		const {outputText} = ts.transpileModule(fixtureContent, {
			fileName: fixtureFile,
			compilerOptions: tsCompilerOptions,
			transformers: {
				before: [
					transformer(),
				],
			},
		});

		expect(outputText).toMatchSnapshot();
	});
})
