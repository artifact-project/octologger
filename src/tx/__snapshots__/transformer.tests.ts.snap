// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformer createProgram 1`] = `
"/Users/k.lebedev/Developer/octologger/src/tx/transformer.fixture.ts

import { logger } from '../logger/logger';
logger.add([\\"src/tx/transformer.fixture.ts\\",3,1], 'ok');
const { info } = logger;
info([\\"src/tx/transformer.fixture.ts\\",6,1], 'wow');
const scope = logger.scope([\\"src/tx/transformer.fixture.ts\\",8,15], 'fetch');
scope.add([\\"src/tx/transformer.fixture.ts\\",10,1], 'start');
scope.done([\\"src/tx/transformer.fixture.ts\\",11,1], 'response');

"
`;

exports[`transformer transpile 1`] = `
"import { logger } from '../logger/logger';
logger.add('ok');
const { info } = logger;
info('wow');
"
`;

exports[`transformer transpileModule 1`] = `
"import { logger } from '../logger/logger';
logger.add('ok');
const { info } = logger;
info('wow');
"
`;
